pkg/test/environment.go

```diff
@@ -0,0 +1,152 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"context"
+	"net"
+
+	"knative.dev/pkg/ptr"
+
+	"github.com/patrickmn/go-cache"
+
+	awscache "github.com/aws/karpenter/pkg/cache"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/fake"
+	"github.com/aws/karpenter/pkg/providers/amifamily"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/providers/instancetype"
+	"github.com/aws/karpenter/pkg/providers/launchtemplate"
+	"github.com/aws/karpenter/pkg/providers/pricing"
+	"github.com/aws/karpenter/pkg/providers/securitygroup"
+	"github.com/aws/karpenter/pkg/providers/subnet"
+
+	coretest "github.com/aws/karpenter-core/pkg/test"
+)
+
+type Environment struct {
+	// API
+	EC2API     *fake.EC2API
+	SSMAPI     *fake.SSMAPI
+	PricingAPI *fake.PricingAPI
+
+	// Cache
+	SSMCache                  *cache.Cache
+	EC2Cache                  *cache.Cache
+	KubernetesVersionCache    *cache.Cache
+	InstanceTypeCache         *cache.Cache
+	UnavailableOfferingsCache *awscache.UnavailableOfferings
+	LaunchTemplateCache       *cache.Cache
+	SubnetCache               *cache.Cache
+	SecurityGroupCache        *cache.Cache
+
+	// Providers
+	InstanceTypesProvider  *instancetype.Provider
+	InstanceProvider       *instance.Provider
+	SubnetProvider         *subnet.Provider
+	SecurityGroupProvider  *securitygroup.Provider
+	PricingProvider        *pricing.Provider
+	AMIProvider            *amifamily.Provider
+	AMIResolver            *amifamily.Resolver
+	LaunchTemplateProvider *launchtemplate.Provider
+	CloudProvider          *cloudprovider.CloudProvider
```

Comment:
I think declaring cloudprovider here is a step too far. I think this is similar to a controller where you should let the user declare this. IMO, this isn't a low-level enough component to declare it here

Reply:
That's fair, I'll update that